src/_pytest/config/__init__.py

```diff
@@ -1656,9 +1662,25 @@ def _getini(self, name: str):
             return _strtobool(str(value).strip())
         elif type == "string":
             return value
+        elif type == "int":
+            if not isinstance(value, (str, int)):
+                raise TypeError(
+                    f"Expected str or int for option {name} of type integer, but got: {value!r}"
+                ) from None
+            return int(value)
+        elif type == "float":
+            if not isinstance(value, (str, float)):
+                raise TypeError(
+                    f"Expected str or flot for option {name} of type float, but got: {value!r}"
+                ) from None
+            return float(value)
         elif type is None:
             return value
         else:
+            if not isinstance(value, (str, Sequence)):
+                raise TypeError(
+                    f"Expected str or sequence for option {name} of type str/list, but got: {value!r}"
+                ) from None
             return self._getini_unknown_type(name, type, value)
```

Comment:
This only raises `ValueError` about the value having an unknown type given. I think we should just adjust `_getini_unknown_type` to accept `value: object` (or `str | list[str] | int | float`, though in this case I don't really see... the value :sweat_smile:).

Reply:
Good catch, done!